### DISCUSSION TITLE
Model Selection - but sharing same covariance / mu

### DISCUSSION THREAD

**Discussion by chrischia06** [skfolio/skfolio#166] [Q&A]
Let's say I wanted to try a bunch of different risk measures, e.g. variance, semivariance, CVAR, and figure out which risk measure works the best (given the same covariance and mu

Is there currently a way to compute MeanRisk() over a different choice of risks or other parameters - but share the same covariance  and mu. calculation so we don't have to repeatedly calculate them?

**Comment by HugoDelatte** [comment]
Great question.

As of `v0.11.0`, this is not yet possible.

To make sure we’re talking about the same thing, let’s take an example:

```python
from sklearn.model_selection import GridSearchCV, train_test_split

from skfolio import RiskMeasure
from skfolio.datasets import load_sp500_dataset
from skfolio.model_selection import WalkForward
from skfolio.moments import LedoitWolf, ShrunkMu
from skfolio.optimization import MeanRisk
from skfolio.preprocessing import prices_to_returns
from skfolio.prior import EmpiricalPrior

prices = load_sp500_dataset()
X = prices_to_returns(prices)

X_train, X_test = train_test_split(X, test_size=0.33, shuffle=False)

ref_model = MeanRisk(
    prior_estimator=EmpiricalPrior(
        mu_estimator = ShrunkMu(),
        covariance_estimator=LedoitWolf(),
    ),
    risk_measure=RiskMeasure.VARIANCE,
)

grid_search = GridSearchCV(
    estimator=ref_model,
    cv=WalkForward(train_size=252, test_size=60),
    n_jobs=-1,
    param_grid={
        "risk_measure": [RiskMeasure.VARIANCE, RiskMeasure.CVAR, RiskMeasure.MAX_DRAWDOWN],
    },
)
grid_search.fit(X_train)
best_model = grid_search.best_estimator_
print(best_model)
```

In this example, `ShrunkMu()` and `LedoitWolf()` are re-fitted for each parameter grid `risk_measure`, even though they don’t change (for a given training fold in the CV).

The reason is that we use the native `GridSearchCV` from scikit-learn, which doesn’t support reusing fitted parts (either in model composition, as in our case, or in Pipelines).

We could potentially implement this in the native scikit-learn `GridSearchCV` or in a dedicated `skfolio` version.
It would partially break parallelization (across parameter grids for a given CV fold), but it’s doable, though it would require clever model comparison logic to determine which components can be safely copied over.

So far, this hasn’t been a problem because these estimations are usually fast compared to the optimization step, so the loss of parallelization may not justify the speed gain of fitted-component copying.

Could you share more details if this feature is critical for you and why? For example, are you using a custom moment estimation that takes significantly longer than the optimization step?

**Reply by chrischia06** [reply]
Thanks - was just curious about it as I saw that the covariance seems to be computed even for other risk measures. 
In my case I have some very large dataframe of asset returns (say T x N is around 300 x  5000) . I also want to calculate it on a rolling basis, 
say [1 to 250, 2 to 251,... to 251 to300]. 

So I was just trying to save as much time as possible.

**Reply by HugoDelatte** [reply]
Thanks for the details.
I see two potential devs:
* Allow the re-use of fitted estimators  in `GridSearchCV`.
* Avoid computing the default moments when they are not used in the next steps.

I will start thinking about the cleanest solution.

**Reply by chrischia06** [reply]
Thanks a lot!
